### ISSUE TITLE
Investigate the addition of an SPI

### ISSUE THREAD

**Issue by arthurdm** [eclipse/microprofile-open-api#212]
The MP OpenAPI spec specifies how to fetch the OpenAPI v3 doc for a single application at `/openapi`.  However, the application server that runs this applications likely has other RESTful APIs that it wants to expose - not from applications, but from the server itself.  Examples include other MicroProfile specs, such as `/health` and `/metrics`, as well as other potential endpoints such as an OpenID Connect endpoint at `/oidc` to manage tokens.

This workitem explores the idea of an SPI that allows vendors to provide OpenAPI v3 definition for these extra "management" side APIs.  

Rough proposal:
(1)  a new Java interface (SPI), where 0..N providers inside the application server can implement:
```
interface OpenAPIProvider
   String getContextRoot()
   String getOpenAPI()
```

(2) a new endpoint, `openapi/providers` which returns a list of currently registered OpenAPIProvider's context roots.

(3) For each `{provider}` in (2), a new URL `openapi/providers/{provider}`.

-----  Example -----

If we have (2) return:
```
'providers':
- metrics
- health
- oidc
```

Then we would have these URLs available in (3):
`openapi/providers/metrics`
`openapi/providers/health`
`openapi/providers/oidc`

**Comment by ivanjunckes** [comment]
+1, I like the proposal.

**Comment by EricWittmann** [comment]
The use-case question is a good one.  The main use case (it seems to me) is that of "discovery" of features provided by the MP platform.  But I think to support that use case we would need to tweak this proposal a bit in one of two ways:

1) Merge all provided OpenAPI documents into one single "platform" or "providers" definition rather than expose each on a separate endpoint.  Perhaps `/openapi/providers` or `/openapi/platform` or something like that.  This would result in two OpenAPI definitions - one for the application and one for the platform.
2) Provide some way to list the providers that are providing an OpenAPI document within `/openapi/providers`.

I prefer (1).

That assumes of course that runtime discovery of provided functionality is an interesting use-case to solve. :)

Perhaps Arthur had something else (or in addition) in mind.

**Comment by arthurdm** [comment]
hey @kenfinnigan / @EricWittmann 

tl;dr - it's about standardizing (via OAS3) discoverability and invocation of platform services

Longer version:

The use case is the discovery of the platform REST endpoints that this MP vendor / app server provides.  These platform endpoints don't necessarily need to be MP specs themselves (eg: OpenID Connect endpoint), but they `can be` (eg: health and metrics).  

So it's about exposing additional services from the platform using OpenAPI v3, instead of having to come up with a way to document these - like using HTML, markdown or other doc patterns.

In the bullet (2) of the proposal it suggests a way to list the providers, so that's borrowing from the `HATEOS` pattern, and illustrates how potential clients of these endpoints can `learn` about these endpoints.

In the future, once we have tool that generates MP Rest Client from an OpenAPI v3 document the usage will be even better:  imagine a developer browsing their favourite MP vendor / app server, picking which endpoints they want to call (can be the app itself or a platform endpoint, eg. get a token from the OpenID Connect endpoint), and then generate a MP Rest Client to invoke that endpoint.

**Comment by kenfinnigan** [comment]
In my view everything you've described as the reason to provide this can be possible, but they're also things that should be secured to prevent malicious use of the endpoints.

If the endpoints are secured the question then becomes whether the OpenAPI doc should also be secured, which I think it does otherwise you're exposing "management" type endpoints.

So if the endpoints and the OpenAPI doc are secured, is an OpenAPI doc really needed?

Right now I'm just not seeing the benefit in making this part of the spec, that's all.

**Comment by arthurdm** [comment]
At the end of the day it's up to the service owner on whether they participate in this SPI and expose their RESTful service as OpenAPI - if they feel that doing so would jeopardize security or just don't feel it's necessary, then they just don't `opt-in` to the SPI.  

In my view, all RESTful services should have an OpenAPI that describes it - afterall, `someone, somewhere` will be consuming that endpoints, right?  =)    If that is made public or not is another question, but I would like the spec to have an opinion on how to expose those in a platform that has services that wish to be exposed.

The security aspect is debatable - we can see many endpoints on the web that expose their Swagger or OpenAPI doc, but to actual invoke it you need an OAuth token or basic auth.  So doc security is not always tied to app security.

**Comment by arthurdm** [comment]
Discussed at the Sept 10 hangout that Arthur will do a PR for this and will go through a final review once ready.

**Comment by arthurdm** [comment]
Delivered in PR #301
